spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/SkipGramFeatureGenerator.java

```diff
@@ -0,0 +1,189 @@
+  private FeatureGeneratorConfig config;
+  private Schema outputSchema;
+  private Word2VecModel loadedModel;
+  private StructType schema;
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+    Preconditions.checkArgument(inputSchema != null, "Input Schema must be a known constant.");
+  }
+
+  @Override
+  public void initialize(SparkExecutionPluginContext context) throws Exception {
+    super.initialize(context);
+    FileSet fileSet = context.getDataset(config.fileSetName);
+    Location modelLocation = fileSet.getBaseLocation().append(config.path);
+    if (!modelLocation.exists()) {
+      throw new IllegalArgumentException(String.format("Failed to find model to use for Regression. Location does " +
+                                                         "not exist: %s.", modelLocation));
+    }
+    // load the model from a file in the model fileset
+    loadedModel = Word2VecModel.load(modelLocation.toURI().getPath());
+  }
+
+  @Override
+  public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext context,
+                                             JavaRDD<StructuredRecord> input) throws Exception {
+    if (input == null) {
+      return context.getSparkContext().emptyRDD();
+    }
+    JavaSparkContext javaSparkContext = context.getSparkContext();
+    SQLContext sqlContext = new SQLContext(javaSparkContext);
+    Schema inputSchema = input.first().getSchema();
+    List<StructField> structField = SparkUtils.getStructFieldList(inputSchema.getFields(),
+                                                                  config.getFeatureListMapping().keySet());
+    schema = new StructType(structField.toArray(new StructField[structField.size()]));
+    JavaRDD<Row> rowRDD = SparkUtils.convertJavaRddStructuredRecordToJavaRddRows(input, schema);
+    DataFrame documentDF = sqlContext.createDataFrame(rowRDD, schema);
+    DataFrame result = null;
+
+    for (Map.Entry<String, String> entry : config.getFeatureListMapping().entrySet()) {
+      result = loadedModel.setInputCol(entry.getKey() + "-transformed").setOutputCol(entry.getValue())
+        .transform(documentDF);
+      documentDF = result;
+    }
+
+    outputSchema = outputSchema != null ? outputSchema : config.getOutputSchema(inputSchema);
+
+    List<Column> requiredFields = new ArrayList<>();
+    for (Schema.Field field : outputSchema.getFields()) {
+      requiredFields.add(new Column(field.getName()));
+    }
+
+    JavaRDD<Row> transformedRDD =
+      javaSparkContext.parallelize(result.select(JavaConversions.asScalaBuffer(requiredFields).toSeq())
+                                     .collectAsList());
+    JavaRDD<StructuredRecord> output = transformedRDD.map(new Function<Row, StructuredRecord>() {
```

Comment:
don't need to convert to DataFrame and all this other weird stuff. Collecting things in the driver should pretty much always be avoided. Basically, if anything other than a function on an RDD is happening, you should reconsider and see if it is really the right way to do it.

This can be done by doing something like:

```

    final Map<String, String> mapping = config.getFeatureListMapping();
    final Splitter splitter = config.getSplitter();
    final int vectorSize = model.wordVectors().length / model.wordIndex().size();
    return input.map(new Function<StructuredRecord, StructuredRecord>() {
      @Override
      public StructuredRecord call(StructuredRecord input) throws Exception {
        StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
        for (Schema.Field field : input.getSchema().getFields()) {
          String fieldName = field.getName();
          builder.set(fieldName, input.get(fieldName));
        }
        for (Map.Entry<String, String> mapEntry : mapping.entrySet()) {
          String inputField = mapEntry.getKey();
          String outputField = mapEntry.getValue();
          String text = input.get(inputField);

          int numWords = 0;
          double[] vectorValues = new double[vectorSize];
          for (String word : splitter.split(text)) {
            double[] wordVector = model.transform(word).toArray();
            for (int i = 0; i < vectorSize; i++) {
              vectorValues[i] += wordVector[i];
            }
            numWords++;
          }
          for (int i = 0; i < vectorSize; i++) {
            vectorValues[i] /= (double) numWords;
          }
          builder.set(outputField, vectorValues);
        }
        return builder.build();
      }
    });
```